draft-irtf-cfrg-opaque.md

```diff
@@ -371,20 +371,32 @@ OPAQUE relies on the following protocols and primitives:
 Note that we only need the base mode variant (as opposed to the verifiable mode
 variant) of the OPRF described in {{I-D.irtf-cfrg-voprf}}.
 
-- Cryptographic hash function:
-  - Hash(m): Compute the cryptographic hash of input message `m`. The type of the
-    hash is determined by the chosen OPRF group.
-  - Nh: The output size of the Hash function.
-
-- Authenticated Key Exchange (AKE, {{instantiations}}):
-  - Npk: The size of the public keys used for the key exchange protocol.
-  - Nsk: The size of the private keys used for the key exchange protocol.
+- Key Derivation Function (KDF):
+  - Extract(salt, ikm): Extract a pseudorandom key of fixed length `Nl` bytes from
+    input keying material `ikm` and an optional byte string `salt`.
+  - Expand(prk, info, L): Expand a pseudorandom key `prk` using optional string `info`
+    into `L` bytes of output keying material.
+  - Nl: The output size of the `Extract()` function in bytes.
+
+- Message Authentication Code (MAC):
+  - MAC(key, msg): Compute a message authentication code over input `msg` with key
+    `key`, producing a fixed-length output of `Nm` bytes.
+  - Nm: The output size of the `MAC()` function in bytes.
+
+- Hash Function:
+  - Hash(msg): Apply a cryptographic hash function to input `msg`, producing an
+    fixed-length digest of size `Nh` bytes.
+  - Nh: The output size of the `Hash()` function in bytes.
 
 - Memory Hard Function (MHF):
   - Harden(msg, params): Repeatedly apply a memory-hard function with parameters
     `params` to strengthen the input `msg` against offline dictionary attacks.
     This function also needs to satisfy collision resistance.
 
+- Authenticated Key Exchange (AKE, {{instantiations}}):
+  - Npk: The size of the public keys used for the key exchange protocol.
+  - Nsk: The size of the private keys used for the key exchange protocol.
```

Comment:
This seems a bit too generic.

The introductory phrase states
> OPAQUE relies on the following protocols and primitives:

So I understand this block is a generic spec not pinning 3DH. It's true we rely on an AKE as a 'Cryptographic Protocol and Algorithm Dependency' here, so maybe we should define the 4 functions to handle 3-way messages? (e.g. start, response, finalize client, finalize server)

Reply:
@bytemare The problem is that there may be future AKEs that support 4-way (or more) messages, so I think even that API can't be specified here...

Reply:
This is borrowed text, but I agree. We can change this to a prime-order group, specifically for 3DH. And the appendix can discuss how this might change if it's needed for other AKEs.